OTAnalytics/plugin_datastore/track_store.py

```diff
@@ -713,39 +766,151 @@ def _get_dataset_with_classes(self, classes: list[str]) -> PandasTrackDataset:
-        return self._filter().get_data()
+    def wrap(self, other: PandasTrackDataset) -> PandasTrackDataset:
+        return FilterByIdPandasTrackDataset(other, self._included_track_ids)
+
+
+class FilterLastNDetectionsPandasTrackDataset(FilteredPandasTrackDataset):
+    def __init__(self, other: PandasTrackDataset, n: int) -> None:
+        super().__init__(other)
+        self._n = n
+        self._cache: PandasTrackDataset | None = None
+
+    def _filter(self) -> PandasTrackDataset:
+        """
+        Filters the track dataset based on included track IDs.
+
+        This method applies a filter on a dataset to include only the tracks
+        specified in `_included_track_ids`. If there are no specified track IDs,
+        it returns the original dataset (`_other`). It employs caching to store the
+        filtered dataset for future use.
+
+        Returns:
+            TrackDataset: The filtered dataset containing only the specified track IDs.
+                If no track IDs are provided, the original dataset is returned.
+        """
+        if self._cache is not None:
+            return self._cache
+
+        logger().info(
+            f"Limit number of detections per track to last {self._n} of them."
+        )
+        filtered_dataset = self._get_dataset_with(last_n=self._n)
+        self._cache = filtered_dataset
+        return filtered_dataset
+
+    def _get_dataset_with(self, last_n: int) -> PandasTrackDataset:
+        if self._other.empty:
+            return self._other
+        dataset = self._other.get_data()
+        filtered_df = get_latest_occurrences(dataset, self._n)
+        # The pandas Index does not implement the Sequence interface, which causes
+        # compatibility issues with the PandasTrackDataset._remove_from_geometry method
+        # when trying to remove geometries for tracks that have been deleted.
+        # To address this, we invalidate the entire geometry cache rather than
+        # attempting selective removal.
+        # This approach is acceptable because track removal only occurs when
+        # cutting tracks, which is a rare use case.
+
+        return PandasTrackDataset(
+            track_geometry_factory=self._other.track_geometry_factory,
+            dataset=filtered_df,
+            geometry_datasets=None,
+            calculator=self._other.calculator,
+        )
+
+    def wrap(self, other: PandasTrackDataset) -> PandasTrackDataset:
+        return FilterLastNDetectionsPandasTrackDataset(other, self._n)
+
+
+def get_latest_occurrences(df: DataFrame, last_n: int) -> DataFrame:
+    index_names = df.index.names
+    df.reset_index(inplace=True)
+    df[RANK] = df.groupby(track.TRACK_ID)[track.OCCURRENCE].rank(
+        method="first", ascending=False
+    )
+    result = df[df[RANK] <= last_n].drop(RANK, axis=1).reset_index(drop=True)
+    return result.set_index(index_names)
+
+
+def get_exactly_two_latest_occurrences_per_id(df: DataFrame, last_n: int) -> DataFrame:
```

Comment:
This is never used. Is this still needed?

Reply:
Thanks, for the hint. The functionality is needed to filter tracks with length 1. Tracks of length 1 exist in live uses cases when the first detection arrives in OTAnalytics.